firstpartydata/first_party_data.go

```diff
@@ -0,0 +1,376 @@
+		if err != nil {
+			errL = append(errL, err)
+		}
+		if len(errL) == 0 {
+			resolvedFpdConfig.Site = newSite
+
+			resolvedFpdData[bName] = resolvedFpdConfig
+		}
+	}
+	return resolvedFpdData, errL
+}
+
+func resolveUser(fpdConfigUser *openrtb2.User, bidRequestUser *openrtb2.User, globalFPD map[string][]byte, openRtbGlobalFPD map[string][]openrtb2.Data, hasGlobalFPD bool, bidderName string) (*openrtb2.User, error) {
+
+	if bidRequestUser == nil && fpdConfigUser == nil {
+		return nil, nil
+	}
+
+	if bidRequestUser == nil && fpdConfigUser != nil {
+		return nil, fmt.Errorf("incorrect First Party Data for bidder %s: User object is not defined in request, but defined in FPD config", bidderName)
+	}
+
+	newUser := *bidRequestUser
+	var err error
+
+	if hasGlobalFPD {
+		//apply global fpd
+		if len(globalFPD[userKey]) > 0 {
+			extData := buildExtData(globalFPD[userKey])
+			if len(newUser.Ext) > 0 {
+				newUser.Ext, err = jsonpatch.MergePatch(newUser.Ext, extData)
+			} else {
+				newUser.Ext = extData
+			}
+		}
+		if len(openRtbGlobalFPD[userDataKey]) > 0 {
+			newUser.Data = openRtbGlobalFPD[userDataKey]
+		}
+	}
+	if fpdConfigUser != nil {
+		//apply bidder specific fpd if present
+		newUser, err = mergeUsers(&newUser, fpdConfigUser)
+	}
+
+	return &newUser, err
+}
+
+func mergeUsers(original *openrtb2.User, fpdConfigUser *openrtb2.User) (openrtb2.User, error) {
+
+	var err error
+	newUser := *original
+	newUser.Keywords = fpdConfigUser.Keywords
+	newUser.Gender = fpdConfigUser.Gender
+	newUser.Yob = fpdConfigUser.Yob
+
+	if len(fpdConfigUser.Ext) > 0 {
+		if len(original.Ext) > 0 {
+			newUser.Ext, err = jsonpatch.MergePatch(original.Ext, fpdConfigUser.Ext)
+		} else {
+			newUser.Ext = fpdConfigUser.Ext
+		}
+	}
+
+	return newUser, err
+}
+
+func resolveSite(fpdConfigSite *openrtb2.Site, bidRequestSite *openrtb2.Site, globalFPD map[string][]byte, openRtbGlobalFPD map[string][]openrtb2.Data, hasGlobalFPD bool, bidderName string) (*openrtb2.Site, error) {
+
+	if bidRequestSite == nil && fpdConfigSite == nil {
+		return nil, nil
+	}
+	if bidRequestSite == nil && fpdConfigSite != nil {
+		return nil, fmt.Errorf("incorrect First Party Data for bidder %s: Site object is not defined in request, but defined in FPD config", bidderName)
+	}
+
+	newSite := *bidRequestSite
+	var err error
+
+	if hasGlobalFPD {
+		//apply global fpd
+		if len(globalFPD[siteKey]) > 0 {
+			extData := buildExtData(globalFPD[siteKey])
+			if len(newSite.Ext) > 0 {
+				newSite.Ext, err = jsonpatch.MergePatch(newSite.Ext, extData)
+			} else {
+				newSite.Ext = extData
+			}
+		}
+		if len(openRtbGlobalFPD[siteContentDataKey]) > 0 {
+			if newSite.Content != nil {
+				contentCopy := *newSite.Content
+				contentCopy.Data = openRtbGlobalFPD[siteContentDataKey]
+				newSite.Content = &contentCopy
+			} else {
+				newSiteContent := &openrtb2.Content{Data: openRtbGlobalFPD[siteContentDataKey]}
+				newSite.Content = newSiteContent
```

Comment:
Style: This could be set directly:
```
newSite.Content = &openrtb2.Content{Data: openRtbGlobalFPD[siteContentDataKey]}
```

.. same for `newAppContent`.

Reply:
Modified